fedimint-server/src/consensus/mod.rs

```diff
@@ -190,205 +186,162 @@ impl FedimintConsensus {
-                        // AcceptedTransactionKey
-                        if !matches!(error, TransactionReplayError(_)) {
-                            dbtx.insert_entry(&RejectedTransactionKey(txid), &format!("{error:?}"))
-                                .await;
-                        }
-                    }
+                    anyhow::bail!("Already received a valid signature share for this peer")
                 }
-            }
-            .instrument(span)
-            .await;
-        }
 
-        rejected_txs
-    }
+                let pks = self.cfg.consensus.auth_pk_set.clone();
 
-    /// Saves the epoch history, calls `end_consensus_epoch` on all modules and
-    /// bans misbehaving peers
-    async fn finalize_process_epoch(
-        &self,
-        dbtx: &mut DatabaseTransaction<'_>,
-        outcome: HbbftConsensusOutcome,
-        rejected_txs: BTreeSet<TransactionId>,
-        consensus_peers: &BTreeSet<PeerId>,
-    ) -> SignedEpochOutcome {
-        let _timing /* logs on drop */ = timing::TimeReporter::new("finalize_process_epoch");
-        let mut drop_peers = Vec::<PeerId>::new();
+                if !pks
+                    .public_key_share(peer_id.to_usize())
+                    .verify(&signature_share.0, self.api.client_cfg.consensus_hash())
+                {
+                    anyhow::bail!("Client config signature share is invalid");
+                }
 
-        self.save_client_config_sig(dbtx, &outcome, &mut drop_peers)
-            .await;
+                // we have received the first valid signature share for this peer
+                dbtx.insert_new_entry(&ClientConfigSignatureShareKey(peer_id), &signature_share)
+                    .await;
 
-        let epoch_history = self
-            .save_epoch_history(outcome.clone(), dbtx, &mut drop_peers, rejected_txs)
-            .await;
+                // collect all valid signature shares received previously
+                let signature_shares = dbtx
+                    .find_by_prefix(&ClientConfigSignatureSharePrefix)
+                    .await
+                    .map(|(key, share)| (key.0.to_usize(), share.0))
+                    .collect::<Vec<_>>()
+                    .await;
 
-        for (module_key, _, module) in self.modules.iter_modules() {
-            let module_drop_peers = module
-                .end_consensus_epoch(consensus_peers, &mut dbtx.with_module_prefix(module_key))
-                .await;
-            drop_peers.extend(module_drop_peers);
-        }
+                if signature_shares.len() <= pks.threshold() {
+                    return Ok(());
+                }
 
-        for peer in drop_peers {
-            dbtx.insert_entry(&DropPeerKey(peer), &()).await;
-        }
+                let threshold_signature = pks
+                    .combine_signatures(signature_shares.iter().map(|(peer, share)| (peer, share)))
+                    .expect("All signature shares are valid");
```

Comment:
Do we really need to abort the process?
```suggestion
                    .context("All signature shares are valid")?;
```

Reply:
We have checked the signature shares before we saved them to disk, so if the validation fails either our cryptography library has a bug and contradicts itself or our database  has been corrupted so I think we should panic.